File: 428-serialize-and-deserialize-n-ary-tree/428-serialize-and-deserialize-n-ary-tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Codec {
    // Encodes a tree to a single string.
    public String serialize(Node root) {
        if(root == null) return "";
        StringBuilder sb = new StringBuilder();
        dfs1(root, sb);
        return sb.toString();
    }
    void dfs1(Node root, StringBuilder sb) {
        if(root == null) {
            sb.append("#,");
            return;
        }
        sb.append(root.val + ",");
        sb.append(root.children.size() + ",");
        for(Node child : root.children) {
            dfs1(child, sb);
        }
    }
	
    // Decodes your encoded data to tree.
    public Node deserialize(String data) {
        if(data == null || data.length() == 0) return null;
        Queue<String> q = new LinkedList<>(Arrays.asList(data.split(",")));
        return dfs2(q);
    }
    Node dfs2(Queue<String> q) {
        String s = q.poll();
        if(s.equals("#")) return null;
        Node node = new Node(Integer.valueOf(s), new ArrayList<>());
        int size = Integer.valueOf(q.poll());
        for(int i = 0; i < size; i++) {
            node.children.add(dfs2(q));
        }
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: 428-serialize-and-deserialize-n-ary-tree/428-serialize-and-deserialize-n-ary-tree.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Codec:
    def serialize(self, root: 'Node') -> str:
        """Encodes a tree to a single string.
        
        :type root: Node
        :rtype: str
        """
        if not root: return None
        str_lst = []
        def dfs(root):
            nonlocal str_lst
            if not root:
                str_lst.append('#')
                return
            str_lst.append(str(root.val))
            str_lst.append(str(len(root.children)))
            for child in root.children:
                dfs(child)
        dfs(root)
        return ','.join(str_lst)
	
    def deserialize(self, data: str) -> 'Node':
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: Node
        """
        if not data: return None
        str_lst = deque(data.split(','))
        def dfs2(str_lst):
            val = int(str_lst.popleft())
            if val == '#':
                return None
            cnt = int(str_lst.popleft())
            root = TreeNode(val)
            root.children = []
            for i in range(cnt):
                root.children.append(dfs2(str_lst))
            return root
        return dfs2(str_lst)
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))